### DISCUSSION TITLE
Using v3 ES modules under jest

### DISCUSSION THREAD

**Discussion by bespokebob** [panva/jose#110] [Q&A]
When using the ES module imports (with typescript + babel), the type checking and build works fine, but running under jest gives me errors like `Cannot find module 'jose/jwe/compact/decrypt'`

I was able to work around by adding this to my jest config:

```javascript
moduleNameMapper: {
    'jose/jwe/compact/decrypt':
      '<rootDir>/node_modules/jose/dist/node/cjs/jwe/compact/decrypt.js',
    'jose/jwe/compact/encrypt':
      '<rootDir>/node_modules/jose/dist/node/cjs/jwe/compact/encrypt.js',
    'jose/jwk/parse':
      '<rootDir>/node_modules/jose/dist/node/cjs/jwk/parse.js',
  },
```

Is there a better solution?

**Comment by panva** [comment]
please see #105. I cannot provide solutions for a deficiency in a test framework that does not respect node runtime require rules and mutates the global namespace on top of that.

If the mapping as-is works for you, cool, but be on a lookout for jest to be properly supporting ES modules and package exports.

**Comment by rachaeldawn** [comment]
Hey, I managed to find a resolution to this. I am using NestJS, and tearing out Jest would've been a _massive_ pain.

I have 2 locations where there is a jest config. 1 in the package.json, and another in the `test/` directory for integration testing.

```json
    "moduleNameMapper": {
      ....
      "jose(.*)": "<rootDir>/../node_modules/jose/dist/node/cjs/$1.js"
    },
```

This may require _slight_ adjustments depending on very specific testing environments (folder structure in particular), but in general, this works just fine and does not require a 1-line-per-file approach.

**Reply by mattvb91** [reply]
Im getting the following error using this:

```
  Configuration error:
    
    Could not locate module jose mapped as:
    /app/node_modules/jose/dist/node/cjs/$1.js.
    
    Please check your configuration for these entries:
    {
      "moduleNameMapper": {
        "/jose(.*)/": "/app/node_modules/jose/dist/node/cjs/$1.js"
      },
      "resolver": undefined
    }

    > 1 | import { jwtVerify } from "jose";
```

any chance anyone can help me with this?

**Comment by panva** [comment]
I'm done waiting for these ecosystems to catch up. jose v4.x will be coming out today/tomorrow with only top level named exports that solve all these tools' problems.
